### PR TITLE
KEYCLOAK-15932 Adding a sample vault configuration

### PR DESCRIPTION
**Pull request by andymunro**
@hmlnarik Can you check if this works?

### CODE REVIEW

**Comment by hmlnarik** [comment]
@sguilhen Could you please check this PR?

File: server_admin/topics/vault.adoc [hunk 1]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
 The configuration above tells {project_name} to setup the custom Elytron provider and use the key resolver that is created by
 the custom factory.
 endif::[]
+
+=== Sample Configuration
+
+The following is an example of configuring a vault and credential store.  The procedure involves two parts:
+
+* Creating the credential store and a vault, where the credential store and vault passwords are in plain text.
+* Updating the credential store and vault to have the password encrypted by using a mask provided by `elytron-tool.sh`.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: The password is actually never encrypted, just masked. I would rephrase this to remove any reference to encryption.

File: server_admin/topics/vault.adoc [hunk 2]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
+
+==== Configuring the credential store and vault without encryption
+
+You create the credential store and a vault where the credential store and vault passwords are in plain text.
+
+.Prerequisites
+
+* A running LDAP instance has `BIND DN credential: secret12`.  
+
+* The alias uses the format <realm-name>_< key-value> when using the default key resolver. In this case, the instance is running in the realm `ldaptest` and `ldaptest_ldap_secret` is the alias that corresponds to the value `ldap_secret` in that realm.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc:237]: the resolver actually replaces any occurence of underscore character (which is used as the separator between realm and key name) with double underscore characters in both realm and key names. That means that if we have, for example, a key named `ldap_secret`, the final combined key will be `ldaptest_ldap__secret` - notice the final key has replaced the original `_` in the key name with double underscore chars.

File: server_admin/topics/vault.adoc [hunk 3]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
+
+.Procedure
+
+. Create the credential store.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: "Create the Elytron credential store"

File: server_admin/topics/vault.adoc [hunk 4]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
++
+[source,bash,subs=+attributes]
+----
+[standalone@localhost:9990 /] /subsystem=elytron/credential-store=test-store:add(create=true, location=/home/test/test-store.p12, credential-reference={clear-text=testpwd1!},implementation-properties={keyStoreType=PKCS12})
+----
+
+. Add an alias to the credential store.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc:250]: It might be good to reinforce somehow that the alias must match what the resolver would produce as the final key - i.e. if using the default resolver the final key is realm_key where occurrences of `_` in both realm and key must be replaced by double`_` . Notice how command line bellow is correctly doing that with the 'ldap_secret' part.

File: server_admin/topics/vault.adoc [hunk 5]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
+----
+
+. Add an alias to the credential store. 
++
+[source,bash,subs=+attributes]
+----
+/subsystem=elytron/credential-store=test-store:add-alias(alias=ldaptest_ldap__secret,secret-value=secret12)
+----
+
+. List the aliases from the credential store.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: This is optional and serves only as a way to visually inspect the contents of the keystore that is produced by Elytron.

File: server_admin/topics/vault.adoc [hunk 6]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
+keytool -list -keystore ./test-store.p12 -storetype PKCS12 -storepass testpwd1!
+Keystore type: PKCS12
+Keystore provider: SUN
+
+Your keystore contains 1 entries
+
+ldaptest_ldap__secret/passwordcredential/clear/, Oct 12, 2020, SecretKeyEntry, 
+----
+
+. Configure the vault SPI.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: "Configure the Vault SPI in Keycloak"

File: server_admin/topics/vault.adoc [hunk 7]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
+[source,bash,subs=+attributes]
+----
+/subsystem=elytron/credential-store=test-store:add-alias(alias=ldaptest_ldap__secret,secret-value=secret12)
+----
+
+. List the aliases from the credential store.
++
+[source,bash,subs=+attributes]
+----
+keytool -list -keystore ./test-store.p12 -storetype PKCS12 -storepass testpwd1!
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: Minor comment: this assumes the keytool command is being run in the /home/test/ folder as that is the full path of the store that we created above. It might be good to change the -keystore param to the full path to avoid any misunderstandings. So I would change it to -keystore /home/test/test-store.p12 in the above command.

File: server_admin/topics/vault.adoc [hunk 8]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
+. Configure the vault SPI. 
++
+[source,bash,subs=+attributes]
+----
+/subsystem=keycloak-server/spi=vault:add(default-provider=elytron-cs-keystore)
+
+/subsystem=keycloak-server/spi=vault/provider=elytron-cs-keystore:add(enabled=true, properties={location=>/home/test/test-store.p12, secret=>testpwd1!, keyStoreType=>PKCS12})
+----
++
+At this point, the vault and credentials store passwords are not encrypted.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: Again, usage of encrypted - I would avoid this sentence. The credential is stored in plaintext form as a SecretKey in the Keystore produced by Elytron, and the password to the store is being used in plaintext form in the Keycloak Vault SPI configuration. What we can do to avoid leaving this password in plaintext is to mask it, but not encrypt it.

File: server_admin/topics/vault.adoc [hunk 9]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
+
+. In the LDAP provider, replace `binDN credential` with `${vault.ldap_secret}`.
+
+. Test your LDAP connection.
++
+.LDAP Vault
+image:images/ldap-vault.png[LDAP Vault]
+
+
+==== Adding encryption to the credential store and vault
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: What we do is we mask only the password that can open the Elytron credential store (which in this case is a Java KeyStore). No encryption involved.

File: server_admin/topics/vault.adoc [hunk 10]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
++
+For example:
++
+[source,bash,subs=+attributes]
+----
+elytron-tool.sh mask --salt 12345678 --iteration 123 --secret testpwd1!
+MASK-3BUbFEyWu0lRAu8.fCqyUk;12345678;123
+----
+
+. Store the masked password on the credential store.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: "Update the Elytron credential store configuration to use the masked password"

File: server_admin/topics/vault.adoc [hunk 11]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
++
+[source,bash,subs=+attributes]
+----
+/subsystem=elytron/credential-store=cs-store:write-attribute(name=credential-reference.clear-text,value="MASK-3BUbFEyWu0lRAu8.fCqyUk;12345678;123")
+----
+
+. Update the vault password.
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc]: "Updated the Keycloak vault configuration to use the masked password"

File: server_admin/topics/vault.adoc [hunk 12]
```diff
@@ -216,3 +216,161 @@ To install and use the above custom provider the configuration would look someth
...
++
+[source,bash,subs=+attributes]
+----
+/subsystem=keycloak-server/spi=vault/provider=elytron-cs-keystore:remove()
+/subsystem=keycloak-server/spi=vault/provider=elytron-cs-keystore:add(enabled=true, properties={location=>/home/test/test-store.p12, secret=>”MASK-3BUbFEyWu0lRAu8.fCqyUk;12345678;123”, keyStoreType=>PKCS12})
```
Comment (sguilhen) [inline server_admin/topics/vault.adoc:346]: This is a valid way, but I think it is possible to update the "properties" param without removing the whole provider using 'write-attribute'. Need to check though.

**Review by sguilhen — CHANGES_REQUESTED** [review]

File: server_admin/topics/vault.adoc (hunk 2 above)
Comment (andymunro) [reply server_admin/topics/vault.adoc:237]: Thanks, @sguilhen.  Are you suggesting that I add this as note, change the example to avoid underscores, or something else.

Comment (sguilhen) [reply server_admin/topics/vault.adoc:237]: @andymunro I would add a note about it, I think it is good that the example contains underscores.

File: server_admin/topics/vault.adoc (hunk 12 above)
Comment (andymunro) [reply server_admin/topics/vault.adoc:346]: Okay, thanks.  Let me know and I can change it.

**Comment by andymunro** [comment]
@sguilhen Thanks for the quick update. Somehow I don't see your new comments.

File: server_admin/topics/vault.adoc (hunk 12 above)
Comment (andymunro) [reply server_admin/topics/vault.adoc:346]: @sguilhen Do you have more comments on this PR?

Comment (sguilhen) [reply server_admin/topics/vault.adoc:346]: @andymunro no, I think it looks good!

**Review by sguilhen — APPROVED** [review]

**Comment by sguilhen** [comment]
It looks good to me now, thanks @andymunro !

**Comment by andymunro** [comment]
@hmlnarik Can you approve and merge this?
